modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/TransportSearchTemplateAction.java

```diff
@@ -71,18 +71,29 @@ public TransportSearchTemplateAction(
     @Override
     protected void doExecute(Task task, SearchTemplateRequest request, ActionListener<SearchTemplateResponse> listener) {
         final SearchTemplateResponse response = new SearchTemplateResponse();
+        boolean success = false;
         try {
             SearchRequest searchRequest = convert(request, response, scriptService, xContentRegistry, searchUsageHolder);
             if (searchRequest != null) {
-                client.search(searchRequest, listener.delegateFailureAndWrap((l, searchResponse) -> {
+                client.search(searchRequest, listener.delegateResponse((l, e) -> {
+                    response.decRef();
+                    l.onFailure(e);
+                }).delegateFailureAndWrap((l, searchResponse) -> {
                     response.setResponse(searchResponse);
-                    l.onResponse(response);
+                    searchResponse.incRef();
```

Comment:
❓  Why do we need to increase the ref explicitly here? Maybe we need SearchTemplateResponse.setResponse() to increase the ref instead of doing so here? 🤔

Reply:
I mainly did it because I handled it the same way for the multi search response. There it made perfect sense because most usage was along the lines of "create the response and put it into the multi-search-response right away". Here it's not quite as clean, but I was thinking of going for a follow-up here anyway to make the template response immutable if possible and then just use the ref count in the search response instead of having its own ref-count. But that takes a couple steps :)